src/orders/views/OrderReturn/OrderReturn.tsx

```diff
@@ -87,11 +76,86 @@ const OrderReturn: React.FC<OrderReturnProps> = ({ orderId }) => {
         },
       }),
     );
+
+    if (returnErrors.length) {
+      return;
+    }
+
+    const grantRefundData = formData.autoGrantRefund
+      ? await grantRefund({
+          variables: {
+            orderId,
+            amount: formData.amount,
+            reason: "",
+            lines: [
+              ...formData.fulfilledItemsQuantities.map(line => ({
+                id: line.data.orderLineId,
+                quantity: line.value,
+              })),
+              ...formData.unfulfilledItemsQuantities.map(({ id, value }) => ({
+                id,
+                quantity: value,
+              })),
+            ],
+            grantRefundForShipping: formData.refundShipmentCosts,
+          },
+        })
+      : null;
+
+    const grantRefundErrors =
+      grantRefundData?.data?.orderGrantRefundCreate?.errors ?? [];
+
+    const grantedRefundId =
+      grantRefundData?.data?.orderGrantRefundCreate?.grantedRefund?.id;
+    const isSendRefund = grantedRefundId && formData.autoSendRefund;
+    const sendRefundErrors = isSendRefund
+      ? await extractMutationErrors(
+          sendRefund({
+            variables: {
+              grantedRefundId,
+              transactionId: data?.order?.transactions[0].id,
+            },
+          }),
+        )
+      : [];
+
+    // If return was successful, but grant/send refund failed, we need to refetch order details
+    // to display updated order data affected by return mutation
+    if (grantRefundErrors.length || sendRefundErrors.length) {
+      refetchOrderDetails();
+    }
+
+    const errors = [...returnErrors, ...grantRefundErrors, ...sendRefundErrors];
+
+    if (errors.some(err => err.code === OrderErrorCode.CANNOT_REFUND)) {
+      notify({
+        autohide: 5000,
+        status: "error",
+        text: intl.formatMessage(messages.cannotRefundDescription),
+        title: intl.formatMessage(messages.cannotRefundTitle),
+      });
+    }
+
+    if (!errors.length) {
+      notify({
+        status: "success",
+        text: intl.formatMessage(
+          getSuccessMessage(formData.autoGrantRefund, isSendRefund),
+        ),
+      });
+      navigate(orderUrl(replacedOrder ?? orderId));
+    }
```

Comment:
these error handlings can be also part of the hook I mentioned - it can return error field along wth eg. some info that given error occurs fe. `error.isCannotRefund` etc.